File: BlockchainJava/BlockchainFundamentals/src/com/blockchain/Block.java
package com.blockchain;

import java.util.Date;
import java.util.List;

/**
 * 
 * @author anubhp
 * 
 * timestamp: date of creation of block.
 * transactions: list of transactions validated and added to the block.
 * nonce: random number to be calculated so that the hash of the the block has some features(here, hash will start from some number of zeros).
 * prevHash: contains the hash of the previous block in the blockchain.
 * hash: contains the hash of the block.
 *
 */
public class Block {
	 private String prevHash;
	 private Date timeStamp;
	 private List<Transaction> transactions;
	 private String hash;
	 private long nonce;
	 
	 public Block(List<Transaction> transactions) {
			this.prevHash = null;
			this.timeStamp = new Date();
			this.transactions = transactions;
			this.nonce = 0;
			this.hash = calculateHash();
		}

	
	 /**
	 * to calculate the hash of the block.
	 * @return hash of the block
	 */
	 public String calculateHash() {
		return CryptoSHA256.getSha256(this.toString());
	}
	
	
	 /**
	  * to mine the block with the given difficulty number.
	  * @param difficulty: here, number of zeros at the start of the hash of every block for the hash to be a valid hash.
	  */
	 public void mineBlock(int difficulty) {
		String difficultyZeroString = "";
		for (int i = 1; i <= difficulty; i++){
			difficultyZeroString += "0";
		}
		this.nonce = 0;
		while (!this.hash.substring(0, difficulty).equals(difficultyZeroString)) {
			this.nonce += 1;
			this.hash = calculateHash();
		}
	}
	
	public String getPrevHash() {
		return prevHash;
	}
	public void setPrevHash(String prevHash) {
		this.prevHash = prevHash;
	}
	public Date getTimeStamp() {
		return timeStamp;
	}
	public void setTimeStamp(Date timeStamp) {
		this.timeStamp = timeStamp;
	}
	public List<Transaction> getTransactions() {
		return transactions;
	}
	public void setTransactions(List<Transaction> transactions) {
		this.transactions = transactions;
	}
	public String getHash() {
		return hash;
	}
	public void setHash(String hash) {
		this.hash = hash;
	}
	public long getNonce() {
		return nonce;
	}
	public void setNonce(long nonce) {
		this.nonce = nonce;
	}
	


	@Override
	public String toString() {
		return "\n\t\tBlock: [\n\t\t\ttimeStamp=" + timeStamp + ", \n\t\t\ttransactions=" + transactions + "\n\t\t\tprevHash=" + prevHash + ", \n\t\t\thash="
				+ hash + ", \n\t\t\tnonce=" + nonce + "\n\t\t]";
	}
	
	
	

}

File: BlockchainCpp/Block.cpp
#include "Block.h"
#include "sha256.h"
#include "utils.h"


using namespace std;
using namespace utils; 

Block::Block(vector<Transaction> transactions) {
    this->prevHash = string();
    this->timeStamp = time(0);
    this->transactions = transactions;
    this->nonce = 0;
    this->hash = this->calculateHash();
}

string Block::getPrevHash() {
	return prevHash;
}
void Block::setPrevHash(string prevHash) {
    this->prevHash = prevHash;
}
time_t Block::getTimeStamp() {
    return timeStamp;
}
void Block::setTimeStamp(time_t timeStamp) {
    this->timeStamp = timeStamp;
}
vector<Transaction> Block::getTransactions() {
    return transactions;
}
void Block::setTransactions(vector<Transaction> transactions) {
    this->transactions = transactions;
}
string Block::getHash() {
    return hash;
}
void Block::setHash(string hash) {
    this->hash = hash;
}
long Block::getNonce() {
    return nonce;
}
void Block::setNonce(long nonce) {
    this->nonce = nonce;
}

string Block::timeToString(time_t timeStamp){
    tm *ltm = localtime(&timeStamp);
    return to_String(1+ltm->tm_hour) + ":" + to_String(1+ltm->tm_min) + ":" + to_String(1+ltm->tm_sec) + "." +
        to_String(ltm->tm_mday) + "-" + to_String(1+ltm->tm_mon) + "-" + to_String(1900+ltm->tm_year);
}

string Block::toString() {
    return "\n\t\tBlock: [\n\t\t\ttimeStamp=" + this->timeToString(this->timeStamp) + ", \n\t\t\ttransactions=" + vectorToString(this->transactions) + "\n\t\t\tprevHash=" + this->prevHash + ", \n\t\t\thash="
				+ this->hash + ", \n\t\t\tnonce=" + to_String(this->nonce) + "\n\t\t]";
}

string Block::calculateHash() {
    return sha256(this->toString());
}

void Block::mineBlock(int difficulty) {
    string difficultyZeroString = "";
		for (int i = 1; i <= difficulty; i++){
			difficultyZeroString += "0";
		}
		this->nonce = 0;
		while (this->hash.substr(0, difficulty).compare(difficultyZeroString) != 0) {
			this->nonce += 1;
			this->hash = calculateHash();
		}

}



